server.js

```diff
@@ -8,6 +8,32 @@ import { callOnChange } from "./api/server/onchange.js";
 import { callOnSubmit } from "./api/server/onsubmit.js";
 import shared from "./shared";
 import log from "./lib/log";
+import { getFunctionParameters } from "./lib/utils";
+
+const handleCallbackFuncParameters = func => {
+  const parameters = getFunctionParameters(func);
+  const handler = {
+    getOwnPropertyDescriptor(target, keyIndex) {
+      const key = keyIndex.split("__-__")[0];
+      const index = parseInt(keyIndex.split("__-__")[1]);
+      if (
+        (key === "game" && index === 0) ||
+        (key === "round" && index === 1) ||
+        (key === "stage" && index === 2)
+      ) {
+        return;
+      } else {
+        throw new Error(`"${key}" property is not allowed on this callback`);
+      }
```

Comment:
I'll be honest with you, I am not 100% sure I can read through all this and know exactly what's going on. 😄 

If I do understand without running, you are checking args as soon as the callback is passed to the engine, which is a great idea (does kinda break people who are not using that arg but still have it in the function's args, but it's better than nothing). And I assume you tested this and it works as expected.

Now, still assuming I understand what's going on, if someone were to pass `players` as the second arg here, it  would not work. That's good. We could go crazy and make sure the right args are given for the right callback, let's not worry about that for now. But it would be nice to help people transition from the previous API, so we should tell people who are still using the `players` arg that they should replace that with `game.players`. So simply a new condition:

```js
else  if (key === "players") {
        throw new Error(`the "players" argument has been deprecated, use "game.players" instead`);
} else {
```

In an ideal world, we would first release this as a deprecation warning, then remove, but we're not big enough of a project with enough resources to go into that lengthly transition, and new projects would not encounter this issue.